### DISCUSSION TITLE
Prebuilds for GitHub Classroom Repos

### DISCUSSION THREAD

**Discussion by mchisolm0** [community/community#34666] [GitHub Education]
Hello,

I am trying to wrap my head around how GitHub Classroom with CodeSpaces works. I love the idea, but when students accept an assignment and try to create a CodeSpace on trunk/main does it not use the prebuild I set up for trunk for that repo? Does it have to build the CodeSpace from scratch for each student?

I ask because I was having students with a lot of issues with long load times and regular failures to build. Granted, I am having the CodeSpace install expo-cli, I just thought it would be able to do that as a prebuild and then just load the image when students start their work. Am I thinking of this wrong? I have not found documentation about prebuilds with GitHub Classroom, so it exists please point me to it. Thank you so much in advance. 

- Matthew Chisolm

**Comment by antonblr** [comment]
Don't have hands on experience with GitHub Classroom / CodeSpaces, but looking at the docs your thoughts are correct -  prebuilds should be built once (and/or on every config change) and then re-used for every started CodeSpaces for that branch.

Docs:
* [Using Codespaces with GitHub Classroom](https://docs.github.com/en/education/manage-coursework-with-github-classroom/integrate-github-classroom-with-an-ide/using-github-codespaces-with-github-classroom)
* [GitHub CodeSpaces overview](https://docs.github.com/en/codespaces/overview)
* [Prebuilding your codespaces](https://docs.github.com/en/codespaces/prebuilding-your-codespaces/about-github-codespaces-prebuilds)
* [Troubleshooting prebuilds](https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-prebuilds)

**Reply by mchisolm0** [reply]
Thank you. I had read those, the part that I am not entirely understanding is the part that I think is specific to how GitHub Classroom works. I create an assignment and set the assignment to use the https://github.com/Mobile-Apps-Chisolm/expo-template-js repo as a template (which I have set to prebuild). Students then navigate to the assignment and use the green "Create codespace on trunk / main" button. 

This is where I start feeling like I don't understand. My students (if the creation succeeds) are presented with the codespace with the above repo, appended with their username or team name for group assignments. While I am still new to GitHub, that is a signal that the codespace is actually a *fork* of the template, right? And that is why the prebuild I have set up doesn't matter?

**Reply by antonblr** [reply]
* Have you tried selecting "Configure and create codespace"? 
<img src="https://user-images.githubusercontent.com/44246099/193394019-3d4c3380-1f6f-4115-a62f-d17d79ff01fb.png" width="350">

* It will present the following options, including the ability to select the "Machine type":
<img src="https://docs.github.com/assets/cb-101608/images/help/codespaces/advanced-options.png" width="500">

* And that drop-down should show if there any `prebuilds` available:
<img src="https://docs.github.com/assets/cb-50286/images/help/codespaces/choose-custom-machine-type.png" width="500">

Though, I'm navigating blind through this, since I don't have access to Classroom / Codespaces setup.

**Reply by mchisolm0** [reply]
That's a good shout, but when I go through that process on one of my student's assignment repos, they do not have prebuilds on the codespace. Again, I think that to make things work behind the scenes, Classroom is actually making a fork for each student or team. That means that they are not actually getting the prebuild setup for the original template repo I have them starting the Codespace from.

Does that make sense?

I think for my use case unless I the Classroom team had a feedback section and were able to add something for the forks to also use the prebuild for the template, I may just need to forgo the automations Classroom gives me.

**Reply by antonblr** [reply]
I think I understand it better now - Classroom will create a new branch for every student for the given repo, but typical Codespace prebuilt is branch-specific (i.e. applies to `main` / `master` / `trunk` branch only). 
Found this interesting phrase [here](https://docs.github.com/en/codespaces/prebuilding-your-codespaces/configuring-prebuilds) : 
>**Any branches** created from a prebuild-enabled parent branch will **typically** also get prebuilds for the same dev container configuration. This is because the prebuild for child branches that use the same dev container configuration as the parent branch are, for the most part, identical, so developers can benefit from faster codespace creation times on those branches also.

The funny part is that I cannot find what does **"typically"** implies.

**Reply by mchisolm0** [reply]
That sounds exactly perfect for my and my students' use case, but it seems when Classroom makes the student branches it is building their workspace from scratch. That is a problem because 1) Since there is a lot to download (e.g. expo-cli etc.) the build takes around 10 min. to build and 2) I have had probably about a...70-80% success rate? The other 20-30% of my students consistently get errors from the build process either saying the build failed because there was "No `CodespaceStateWatcher` found" or 404 error, but those are separate issues that (~~I think~~ I did) I made a separate discussion for.

link to other discussion for reference: https://github.com/orgs/community/discussions/34738

**Reply by mchisolm0** [reply]
Looking back at my template repo, its looking like Classroom actually makes *forks* for my students. 
<img width="1305" alt="image" src="https://user-images.githubusercontent.com/39521893/193455415-5375608c-672f-42bd-84de-71fc6a202853.png">

I will see if there is an option in Classroom to default to making branches instead, but it is looking more and more like I won't be able to use Github Classroom.

**Reply by mchisolm0** [reply]
Actually...If my last post that it was making forks for my students was right, would I not see those forks here?
<img width="1317" alt="image" src="https://user-images.githubusercontent.com/39521893/193455731-9cbf029c-f6bb-48e8-a7ba-dd7cd7122e52.png">

**Comment by mchisolm0** [comment]
I am going back through the process of creating an assignment and documenting it here for anyone able to help who is unfamiliar with Github Classroom.

1. Create a new assignment 
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/39521893/193455951-f0bfa2fd-824b-4121-913c-c9f5a81f545b.png">

2. Title assignment, set public/private (the assignment in this post was set to private and I'm making a new one as public to see if that changes the problem), and give admin permissions 
<img width="1020" alt="image" src="https://user-images.githubusercontent.com/39521893/193456135-cf059be0-225d-42ef-84c0-083117d13eb5.png">

3. Choose template repo and IDE. I am thinking the fact that it uses a template repo might also be the source of the issue. 
<img width="1029" alt="image" src="https://user-images.githubusercontent.com/39521893/193456243-73fc44a3-894f-4ccb-b7ad-ae10bddbd422.png">

4. Add tests and a PR for feedback. I think this is very unlikely related to the issue 
<img width="1033" alt="image" src="https://user-images.githubusercontent.com/39521893/193456289-e909244f-6046-494b-b3b6-5374394a4820.png">

Let me know if you have any questions.

**Comment by ossanna16** [comment]
Great question @mchisolm0! I'm glad you found our community 🙂

**Reply by mchisolm0** [reply]
Thanks, I am, too 😁

**Comment by aschilling** [comment]
@mchisolm0 Thanks for the issue you raised. I am currently facing the same problem. Did you find a workaround for prebuilding?

**Reply by mchisolm0** [reply]
No, the only thing I have thought of is once students have accepted an assignment having them use GitPod.io which is similar to CodeSpaces but a different company. It is a little bit easier for students to set up prebuilds, but it still means we are having to set them up manually after the students accept the assignment.

I may look at it some more over the holiday break. I just have not had the time to focus on this since I posted the original issue.

**Reply by mchisolm0** [reply]
I will say I one of the last thoughts I had when I was working on this was to create a GitHub action. There is a way to create a branch, but I could not find any documentation or actions in the marketplace that could retrieve the teams and their members. Here is a link to the post I did, though, it currently has no answers. https://github.com/orgs/community/discussions/34947

Also, putting fresh eyes on that approach, I am unconvinced it would help our use-case.
